File: Source/OldCode/Archive/2004 TVA Code Library/Source/EE/Phasor/FrequencyDefinitionBase.vb
'***********************************************************************
'  FrequencyDefinitionBase.vb - Frequency and df/dt value definition base class
'  Copyright © 2005 - TVA, all rights reserved
'
'  Build Environment: VB.NET, Visual Studio 2003
'  Primary Developer: James R Carroll, System Analyst [TVA]
'      Office: COO - TRNS/PWR ELEC SYS O, CHATTANOOGA, TN - MR 2W-C
'       Phone: 423/751-2827
'       Email: jrcarrol@tva.gov
'
'  Code Modification History:
'  ---------------------------------------------------------------------
'  02/18/2005 - James R Carroll
'       Initial version of source generated
'
'***********************************************************************

Namespace EE.Phasor

    ' This class represents the common implementation of the protocol independent definition of a frequency and df/dt value.
    Public MustInherit Class FrequencyDefinitionBase

        Inherits ChannelDefinitionBase
        Implements IFrequencyDefinition

        Private m_dfdtScale As Integer
        Private m_dfdtOffset As Double

        Protected Sub New(ByVal parent As IConfigurationCell)

            MyBase.New(parent)

            m_dfdtScale = 1

        End Sub

        Protected Sub New(ByVal parent As IConfigurationCell, ByVal dataFormat As DataFormat, ByVal index As Integer, ByVal label As String, ByVal scale As Integer, ByVal offset As Double, ByVal dfdtScale As Double, ByVal dfdtOffset As Double)

            MyBase.New(parent, dataFormat, index, label, scale, offset)

            Me.DfDtScalingFactor = dfdtScale
            m_dfdtOffset = dfdtOffset

        End Sub

        ' Dervied classes are expected to expose a Public Sub New(ByVal frequencyDefinition As IFrequencyDefinition)
        Protected Sub New(ByVal frequencyDefinition As IFrequencyDefinition)

            Me.New(frequencyDefinition.Parent, frequencyDefinition.DataFormat, frequencyDefinition.Index, frequencyDefinition.Label, frequencyDefinition.ScalingFactor, _
                frequencyDefinition.Offset, frequencyDefinition.DfDtScalingFactor, frequencyDefinition.DfDtOffset)

        End Sub

        Public ReadOnly Property NominalFrequency() As LineFrequency Implements IFrequencyDefinition.NominalFrequency
            Get
                Return Parent.Parent.NominalFrequency
            End Get
        End Property

        Public Overridable Property DfDtOffset() As Double Implements IFrequencyDefinition.DfDtOffset
            Get
                Return m_dfdtOffset
            End Get
            Set(ByVal Value As Double)
                m_dfdtOffset = Value
            End Set
        End Property

        Public Overridable Property DfDtScalingFactor() As Integer Implements IFrequencyDefinition.DfDtScalingFactor
            Get
                Return m_dfdtScale
            End Get
            Set(ByVal Value As Integer)
                If Value > MaximumDfDtScalingFactor Then Throw New OverflowException("DfDt scaling factor value cannot exceed " & MaximumDfDtScalingFactor)
                m_dfdtScale = Value
            End Set
        End Property

        Public Overridable ReadOnly Property MaximumDfDtScalingFactor() As Integer Implements IFrequencyDefinition.MaximumDfDtScalingFactor
            Get
                ' Typical scaling/conversion factors should fit within 3 bytes (i.e., 24 bits) of space
                Return 2 ^ 24
            End Get
        End Property

    End Class

End Namespace

File: Source/Libraries/GSF.PhasorProtocols/FrequencyDefinitionBase.cs
//******************************************************************************************************
//  FrequencyDefinitionBase.cs - Gbtc
//
//  Copyright © 2012, Grid Protection Alliance.  All Rights Reserved.
//
//  Licensed to the Grid Protection Alliance (GPA) under one or more contributor license agreements. See
//  the NOTICE file distributed with this work for additional information regarding copyright ownership.
//  The GPA licenses this file to you under the MIT License (MIT), the "License"; you may
//  not use this file except in compliance with the License. You may obtain a copy of the License at:
//
//      http://www.opensource.org/licenses/MIT
//
//  Unless agreed to in writing, the subject software distributed under the License is distributed on an
//  "AS-IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. Refer to the
//  License for the specific language governing permissions and limitations.
//
//  Code Modification History:
//  ----------------------------------------------------------------------------------------------------
//  02/18/2005 - J. Ritchie Carroll
//       Generated original version of source code.
//  09/15/2009 - Stephen C. Wills
//       Added new header and license agreement.
//  10/5/2012 - Gavin E. Holden
//       Added new header and license agreement.
//  12/17/2012 - Starlynn Danyelle Gilliam
//       Modified Header.
//
//******************************************************************************************************

using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Runtime.Serialization;
using GSF.Units.EE;

namespace GSF.PhasorProtocols
{
    /// <summary>
    /// Represents the common implementation of the protocol independent definition of a <see cref="IFrequencyValue"/>.
    /// </summary>
    [Serializable]
    public abstract class FrequencyDefinitionBase : ChannelDefinitionBase, IFrequencyDefinition
    {
        #region [ Members ]

        // Fields
        private uint m_dfdtScale;
        private double m_dfdtOffset;

        #endregion

        #region [ Constructors ]

        /// <summary>
        /// Creates a new <see cref="FrequencyDefinitionBase"/>.
        /// </summary>
        /// <param name="parent">The <see cref="IConfigurationCell"/> parent of this <see cref="FrequencyDefinitionBase"/>.</param>
        protected FrequencyDefinitionBase(IConfigurationCell parent)
            : base(parent)
        {
        }

        /// <summary>
        /// Creates a new <see cref="FrequencyDefinitionBase"/> from specified parameters.
        /// </summary>
        /// <param name="parent">The <see cref="IConfigurationCell"/> parent of this <see cref="FrequencyDefinitionBase"/>.</param>
        /// <param name="label">The label of this <see cref="FrequencyDefinitionBase"/>.</param>
        /// <param name="scale">The integer scaling value of this <see cref="FrequencyDefinitionBase"/>.</param>
        /// <param name="dfdtScale">The df/dt scaling value of this <see cref="FrequencyDefinitionBase"/>.</param>
        /// <param name="dfdtOffset">The df/dt offset of this <see cref="FrequencyDefinitionBase"/>.</param>
        protected FrequencyDefinitionBase(IConfigurationCell parent, string label, uint scale, uint dfdtScale, double dfdtOffset)
            : base(parent, label, scale, 0.0D)
        {
            m_dfdtScale = dfdtScale;
            m_dfdtOffset = dfdtOffset;
        }

        /// <summary>
        /// Creates a new <see cref="FrequencyDefinitionBase"/> from serialization parameters.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> with populated with data.</param>
        /// <param name="context">The source <see cref="StreamingContext"/> for this deserialization.</param>
        protected FrequencyDefinitionBase(SerializationInfo info, StreamingContext context)
            : base(info, context)
        {
            // Deserialize frequency definition
            m_dfdtScale = info.GetUInt32("dfdtScale");
            m_dfdtOffset = info.GetDouble("dfdtOffset");
        }

        #endregion

        #region [ Properties ]

        /// <summary>
        /// Gets the <see cref="GSF.PhasorProtocols.DataFormat"/> of this <see cref="FrequencyDefinitionBase"/>.
        /// </summary>
        public override DataFormat DataFormat => Parent.FrequencyDataFormat;

        /// <summary>
        /// Gets the nominal <see cref="LineFrequency"/> of this <see cref="FrequencyDefinitionBase"/>.
        /// </summary>
        /// <remarks>
        /// Value returned is the <see cref="IConfigurationCell.NominalFrequency"/> and is exposed here just for convenience.
        /// </remarks>
        public virtual LineFrequency NominalFrequency => Parent.NominalFrequency;

        /// <summary>
        /// Gets or sets the index of this <see cref="FrequencyDefinitionBase"/>.
        /// </summary>
        /// <remarks>
        /// Phasor protocols only define one frequency measurement per device.
        /// </remarks>
        [EditorBrowsable(EditorBrowsableState.Never)]
        public override int Index
        {
            get => base.Index;
            set => base.Index = value;
        }

        /// <summary>
        /// Gets or sets the offset of this <see cref="FrequencyDefinitionBase"/>.
        /// </summary>
        /// <remarks>
        /// Offset for frequency values will always be the nominal frequency as defined in parent configuration cell; assigning a value is not supported.
        /// </remarks>
        /// <exception cref="NotSupportedException">Frequency offset is read-only; value is determined by nominal frequency specified in containing configuration cell.</exception>
        public override double Offset
        {
            get => (double)Parent.NominalFrequency;
            set => throw new NotSupportedException("Frequency offset is read-only; value is determined by nominal frequency specified in containing configuration cell");
        }

        /// <summary>
        /// Gets or sets the df/dt offset of this <see cref="FrequencyDefinitionBase"/>.
        /// </summary>
        public virtual double DfDtOffset
        {
            get => m_dfdtOffset;
            set => m_dfdtOffset = value;
        }

        /// <summary>
        /// Gets or sets the df/dt scaling value of this <see cref="FrequencyDefinitionBase"/>.
        /// </summary>
        public virtual uint DfDtScalingValue
        {
            get => m_dfdtScale;
            set => m_dfdtScale = value;
        }

        /// <summary>
        /// Gets a <see cref="Dictionary{TKey,TValue}"/> of string based property names and values for this <see cref="FrequencyDefinitionBase"/> object.
        /// </summary>
        public override Dictionary<string, string> Attributes
        {
            get
            {
                Dictionary<string, string> baseAttributes = base.Attributes;

                baseAttributes.Add("df/dt Offset", DfDtOffset.ToString());
                baseAttributes.Add("df/dt Scaling Value", DfDtScalingValue.ToString());

                return baseAttributes;
            }
        }

        #endregion

        #region [ Methods ]

        /// <summary>
        /// Populates a <see cref="SerializationInfo"/> with the data needed to serialize the target object.
        /// </summary>
        /// <param name="info">The <see cref="SerializationInfo"/> to populate with data.</param>
        /// <param name="context">The destination <see cref="StreamingContext"/> for this serialization.</param>
        public override void GetObjectData(SerializationInfo info, StreamingContext context)
        {
            base.GetObjectData(info, context);

            // Serialize frequency definition
            info.AddValue("dfdtScale", m_dfdtScale);
            info.AddValue("dfdtOffset", m_dfdtOffset);
        }

        #endregion
    }
}